pytorch_lightning/core/lightning.py

```diff
@@ -1166,16 +1166,44 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
 
         Override for your own behavior
 
+        It works with `untoggle_optimizer` to make sure param_requires_grad_state is properly reset.
+
         Args:
             optimizer:
             optimizer_idx:
         """
-        for param in self.parameters():
-            param.requires_grad = False
+        param_requires_grad_state = {}
+        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
+            if optimizer_idx != opt_idx:
+                for group in opt.param_groups:
+                    for param in group['params']:
+                        param_requires_grad_state[param] = param.requires_grad
+                        param.requires_grad = False
 
         for group in optimizer.param_groups:
             for param in group['params']:
-                param.requires_grad = True
+                if param in param_requires_grad_state:
+                    param.requires_grad = param_requires_grad_state[param]
```

Comment:
Clean-up :broom: 

```suggestion
        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
            for group in opt.param_groups:
                for param in group['params']:
                    if optimizer_idx == opt_idx:
                        param.requires_grad = param_requires_grad_state.get(param, True)
                    else:
                        param.requires_grad = False
                        param_requires_grad_state[param] = param.requires_grad
```

Reply:
This won't work, but we will by sorting optimizer. Good suggestion ! I will do so !

Reply:
@carmocca  There seems to be an issue in the suggested edit for the line
`param.requires_grad = param_requires_grad_state.get(param, True)` 
This would set the `requires_grad` True for all the params of the current optimizer. That's exactly what we want to avoid here.

Reply:
Done !